# Data Prepocessing

In [4]:
import numpy as np
import pandas as pd
import csv
from sklearn import preprocessing
from sklearn.decomposition import PCA

## Bank Dataset

In [104]:
# Read data from file

with open('bank/bank-full.csv', 'r') as csvfile:
    spamreader = csv.reader(csvfile, delimiter=';')
    df = list(spamreader)
    df = np.asarray(df)
    df = pd.DataFrame(data=df[1:], columns=df[0])
    

In [105]:
# Inspecting the unknown values in the dataset

jobs = df['job']
unknown_jobs = 0
for job in jobs:
    if job == 'unknown':
        unknown_jobs = unknown_jobs + 1
        
educations = df['education']
unknown_education = 0
for education in educations:
    if education == 'unknown':
        unknown_education = unknown_education + 1

defaults = df['default']
unknown_default = 0
for default in defaults:
    if default == 'unknown':
        unknown_default = unknown_default + 1

contacts = df['contact']
unknown_contact = 0
for contact in contacts:
    if contact == 'unknown':
        unknown_contact = unknown_contact + 1
    
pdays = df['pdays']
unknown_pday = 0
for pday in pdays:
    if pday == '-1':      # Treating never contacted as unknown
        unknown_pday = unknown_pday + 1

pouts = df['poutcome']
unknown_pouts = 0
for pout in pouts:
    if pout == 'unknown':
        unknown_pouts = unknown_pouts + 1
    

print("Unknown values\n")
print("Unknown jobs :", unknown_jobs)
print("Unknown education :", unknown_education)
print("Unknown contacts :", unknown_contact)
print("Unknown pdays :", unknown_pday)
print("Unknown pouts :", unknown_pouts)



Unknown values

Unknown jobs : 288
Unknown education : 1857
Unknown contacts : 13020
Unknown pdays : 36954
Unknown pouts : 36959


### Observations from Data Set

- Total number of rows: 45211
- Total number of attributes: 16 + output attribute (y)
- Yes samples - 5289
- No samples - 39922
- The number of yes samples are much smaller than the number of no samples indicating that the data set is highly skewed.

- Unknown values :-
    - job - 288
    - education - 1857
    - contact - 13020
    - pdays - 36954
    - poutcome - 36959
- A very large number of rows have pdays = -1 and poutcome = unknown. These represent clients that have not been contacted in the past

In [106]:
df = df.drop('duration', axis=1)             # Removing duration attribute

# Separating columns as numerical and categorical
numeric = ['age', 'balance', 'campaign', 'pdays', 'previous']
categorical = ['job', 'marital', 'education', 'default', 'housing', 'loan', 'contact', 'day', 'month', 'poutcome']


In [107]:
# Substituting pdays = 999 for pdays = -1
pdays = df['pdays']
pdays = pdays.replace('-1', '999')
df['pdays'] = pdays


# Substituting pdays = 999 and poutcome = failure for pdays = -1 and poutcome = unknown
# prev_list = ['pdays', 'poutcome']
# prev_df = df[prev_list]
# prev_df = prev_df.replace({'pdays':'-1', 'poutcome':'unknown'}, {'pdays':'999', 'poutcome':'failure'})
# df[prev_list] = prev_df


In [108]:
# Taking care of categorical values

# factorize assigns unique numeric values to each category
#for c in categorical:
#    df[c] = pd.factorize(df[c])[0]
    
# Use dummy attributes for categorical values
df = pd.get_dummies(df, columns=categorical, prefix=categorical)
df['y'],un = pd.factorize(df['y'])

#df[numeric] = df[numeric].astype('float')
#df[categorical] = df[categorical].astype('category')

df['y'] = df['y'].astype(int) # Converting otuput variable to integers


In [109]:
# Normalizing numerical values

scaler = preprocessing.StandardScaler()
rescaled_df = scaler.fit_transform(df[numeric])
df[numeric] = pd.DataFrame(rescaled_df, columns=numeric)

e:\python\lib\site-packages\sklearn\preprocessing\data.py:617: DataConversionWarning: Data with input dtype object were all converted to float64 by StandardScaler.
  return self.partial_fit(X, y)
e:\python\lib\site-packages\sklearn\base.py:462: DataConversionWarning: Data with input dtype object were all converted to float64 by StandardScaler.
  return self.fit(X, **fit_params).transform(X)


In [62]:
#Saving as pickle file

df.to_pickle('data/bank_data_dummies.pkl') # using dummies and replacing pdays only

In [63]:
# Principal Component Analysis

y_values = df.loc[:, ['y']].values
df.drop(columns=['y'], axis=1, inplace=True)

pca = PCA()
#pca = PCA(n_components=10)
components = pca.fit_transform(df)
var_ratios = pca.explained_variance_ratio_
#rint(var_ratios)

# Calculating the number of attributes for 80% variance
s=0
count = 0
for ratio in var_ratios:
    s = s + ratio
    count = count + 1
    if s >= 0.8:
        break
print(count)

# Calculating the number of attributes for 95% variance
s=0
count1 = 0
for ratio in var_ratios:
    s = s + ratio
    count1 = count1+ 1
    if s >= 0.95:
        break
print(count1)

#df = pd.DataFrame(data=components)
df['y'] = y_values

15
44


### Observations from PCA

- There were 15 attributes (after removing duration) subjected to PCA
- 80% of the variance can be captured by 2 attributes after PCA
- 95% of the variance can be captured by 5 attributes after PCA

In [64]:
df


age   balance  campaign     pdays  previous  job_admin.  \
0      1.606965  0.256419 -0.569351  0.466408 -0.251940           0   
1      0.288529 -0.437895 -0.569351  0.466408 -0.251940           0   
2     -0.747384 -0.446762 -0.569351  0.466408 -0.251940           0   
3      0.571051  0.047205 -0.569351  0.466408 -0.251940           0   
4     -0.747384 -0.447091 -0.569351  0.466408 -0.251940           0   
5     -0.559037 -0.371551 -0.569351  0.466408 -0.251940           0   
6     -1.218254 -0.300608 -0.569351  0.466408 -0.251940           0   
7      0.100181 -0.446762 -0.569351  0.466408 -0.251940           0   
8      1.606965 -0.407679 -0.569351  0.466408 -0.251940           0   
9      0.194355 -0.252657 -0.569351  0.466408 -0.251940           0   
10     0.006007 -0.358742 -0.569351  0.466408 -0.251940           1   
11    -1.124080 -0.319329 -0.569351  0.466408 -0.251940           1   
12     1.136095 -0.445449 -0.569351  0.466408 -0.251940           0   
13     1.606965 -0.424100 -0.569351  0.466408 -0.251940           0   
14     1.512791 -0.394213 -0.569351  0.466408 -0.251940           0   
15     0.947747 -0.372207 -0.569351  0.466408 -0.251940           0   
16     0.382703 -0.443150 -0.569351  0.466408 -0.251940           1   
17     1.512791 -0.430341 -0.569351  0.466408 -0.251940           0   
18     1.795313 -0.427713 -0.569351  0.466408 -0.251940           0   
19    -0.747384 -0.447419 -0.569351  0.466408 -0.251940           0   
20    -1.218254 -0.209960 -0.569351  0.466408 -0.251940           0   
21     1.418617 -0.191568 -0.569351  0.466408 -0.251940           0   
22    -0.841558 -0.439865 -0.569351  0.466408 -0.251940           0   
23    -1.500776 -0.430998 -0.569351  0.466408 -0.251940           0   
24    -0.088167 -0.447419 -0.569351  0.466408 -0.251940           0   
25     0.288529 -0.569598 -0.569351  0.466408 -0.251940           1   
26    -0.182341 -0.363668 -0.569351  0.466408 -0.251940           0   
27     1.041921 -0.410306 -0.569351  0.466408 -0.251940           0   
28     0.476877 -0.528215 -0.246560  0.466408 -0.251940           0   
29    -0.464863 -0.360384 -0.569351  0.466408 -0.251940           0   
...         ...       ...       ...       ...       ...         ...   
45181  0.476877  1.811893 -0.246560 -2.438859  1.050473           0   
45182 -0.653211 -0.403737 -0.246560 -2.211318  1.918749           0   
45183  2.737053 -0.341006 -0.569351 -2.511408  2.787025           0   
45184  2.077835  0.043593 -0.569351 -2.755437  1.918749           0   
45185  1.795313  0.950405 -0.569351 -2.524599  1.484611           0   
45186  1.701139  0.045235 -0.569351 -2.485027  0.616335           0   
45187 -0.841558 -0.063806 -0.569351  0.466408 -0.251940           0   
45188 -1.124080 -0.216529 -0.569351 -1.014256  0.616335           0   
45189 -1.500776 -0.382061 -0.569351 -2.524599  1.918749           0   
45190 -0.841558 -0.402752 -0.569351 -2.208020  1.050473           0   
45191  3.207923  0.803922 -0.569351 -2.224509  0.182198           0   
45192 -1.124080 -0.196166 -0.569351  0.466408 -0.251940           0   
45193 -1.218254 -0.395198 -0.246560 -2.719163  1.484611           0   
45194  1.701139 -0.402095 -0.246560 -2.211318  1.918749           0   
45195  2.548705 -0.071032 -0.569351 -2.211318  2.352887           0   
45196 -1.500776 -0.329839 -0.569351  0.466408 -0.251940           0   
45197 -0.464863  0.048848 -0.569351  0.466408 -0.251940           0   
45198 -0.370689  0.021587 -0.246560  0.466408 -0.251940           0   
45199 -0.653211  0.037024  0.076230 -1.080210  4.957714           0   
45200 -0.276515 -0.264480  0.399020  0.466408 -0.251940           0   
45201  1.136095 -0.255941 -0.569351 -2.221211  1.484611           0   
45202 -0.653211 -0.264480 -0.569351  0.466408 -0.251940           1   
45203 -1.689124 -0.410306 -0.569351  0.466408 -0.251940           0   
45204  3.019575  0.488624 -0.569351 -2.696079  3.221163           0   
45205 -1.500776 -0.281559 -0.246560 

In [65]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45211 entries, 0 to 45210
Data columns (total 81 columns):
age                    45211 non-null float64
balance                45211 non-null float64
campaign               45211 non-null float64
pdays                  45211 non-null float64
previous               45211 non-null float64
job_admin.             45211 non-null uint8
job_blue-collar        45211 non-null uint8
job_entrepreneur       45211 non-null uint8
job_housemaid          45211 non-null uint8
job_management         45211 non-null uint8
job_retired            45211 non-null uint8
job_self-employed      45211 non-null uint8
job_services           45211 non-null uint8
job_student            45211 non-null uint8
job_technician         45211 non-null uint8
job_unemployed         45211 non-null uint8
job_unknown            45211 non-null uint8
marital_divorced       45211 non-null uint8
marital_married        45211 non-null uint8
marital_single         45211 non-null uint8
e

In [66]:
df.describe()

age       balance      campaign         pdays      previous  \
count  4.521100e+04  4.521100e+04  4.521100e+04  4.521100e+04  4.521100e+04   
mean   2.112250e-16  1.760208e-17  3.017500e-17 -8.046667e-17  4.023334e-17   
std    1.000011e+00  1.000011e+00  1.000011e+00  1.000011e+00  1.000011e+00   
min   -2.159994e+00 -3.081149e+00 -5.693506e-01 -2.824689e+00 -2.519404e-01   
25%   -7.473845e-01 -4.237719e-01 -5.693506e-01  4.664082e-01 -2.519404e-01   
50%   -1.823406e-01 -3.002800e-01 -2.465603e-01  4.664082e-01 -2.519404e-01   
75%    6.652252e-01  2.158743e-02  7.622994e-02  4.664082e-01 -2.519404e-01   
max    5.091402e+00  3.309478e+01  1.944365e+01  4.664082e-01  1.191360e+02   

         job_admin.  job_blue-collar  job_entrepreneur  job_housemaid  \
count  45211.000000     45211.000000      45211.000000   45211.000000   
mean       0.114375         0.215257          0.032890       0.027427   
std        0.318269         0.411005          0.178351       0.163326   
min        0.000000         0.000000          0.000000       0.000000   
25%        0.000000         0.000000          0.000000       0.000000   
50%        0.000000         0.000000          0.000000       0.000000   
75%        0.000000         0.000000          0.000000       0.000000   
max        1.000000         1.000000          1.000000       1.000000   

       job_management      ...          month_mar     month_may     month_nov  \
count    45211.000000      ...       45211.000000  45211.000000  45211.000000   
mean         0.209197      ...           0.010551      0.304483      0.087810   
std          0.406740      ...           0.102174      0.460193      0.283022   
min          0.000000      ...           0.000000      0.000000      0.000000   
25%          0.000000      ...           0.000000      0.000000      0.000000   
50%          0.000000      ...           0.000000      0.000000      0.000000   
75%          0.000000      ...           0.000000      1.000000      0.000000   
max          1.000000      ...           1.000000      1.000000      1.000000   

          month_oct     month_sep  poutcome_failure  poutcome_other  \
count  45211.000000  45211.000000      45211.000000    45211.000000   
mean       0.016323      0.012807          0.108403        0.040698   
std        0.126718      0.112441          0.310892        0.197592   
min        0.000000      0.000000          0.000000        0.000000   
25%        0.000000      0.000000          0.000000        0.000000   
50%        0.000000      0.000000          0.000000        0.000000   
75%        0.000000      0.000000          0.000000        0.000000   
max        1.000000      1.000000          1.000000        1.000000   

       poutcome_success  poutcome_unknown             y  
count      45211.000000      45211.000000  45211.000000  
mean           0.033421          0.817478      0.116985  
std            0.179735          0.386278      0.321406  
min            0.000000          0.000000      0.000000  
25%            0.000000          1.000000      0.000000  
50%            0.000000          1.000000      0.000000  
75%            0.000000          1.000000      0.000000  
max            1.000000          1.000000      1.000000  

[8 rows x 81 columns]

In [67]:
#Saving as pickle file

df.to_pickle('data/bank_data_dummies_pca.pkl')    # using dummies and replacing pdays only and performing PCA


### Bank Additional Dataset

In [5]:
# Reading data from file

with open('bank-additional/bank-additional-full.csv', 'r') as csvfile:
    spamreader = csv.reader(csvfile, delimiter=';')
    df = list(spamreader)
    df = np.asarray(df)
    df = pd.DataFrame(data=df[1:], columns=df[0])

In [6]:
# Inspecting the unknown values in the dataset

jobs = df['job']
unknown_jobs = 0
for job in jobs:
    if job == 'unknown':
        unknown_jobs = unknown_jobs + 1
        
maritals = df['marital']
unknown_maritals = 0
for marital in maritals:
    if marital == 'unknown':
        unknown_maritals = unknown_maritals + 1

educations = df['education']
unknown_education = 0
for education in educations:
    if education == 'unknown':
        unknown_education = unknown_education + 1

defaults = df['default']
unknown_default = 0
for default in defaults:
    if default == 'unknown':
        unknown_default = unknown_default + 1

housing = df['housing']
unknown_housing = 0
for house in housing:
    if house == 'unknown':
        unknown_housing = unknown_housing + 1
    
loans = df['loan']
unknown_loans = 0
for loan in loans:
    if loan == 'unknown':
        unknown_loans = unknown_loans + 1
    
pdays = df['pdays']
unknown_pday = 0
for pday in pdays:
    if pday == '999':      # Treating never contacted as unknown
        unknown_pday = unknown_pday + 1

pouts = df['poutcome']
unknown_pouts = 0
for pout in pouts:
    if pout == 'nonexistent':
        unknown_pouts = unknown_pouts + 1
    

print("Unknown values\n")
print("Unknown jobs :", unknown_jobs)
print("Unknown marital status :", unknown_maritals)
print("Unknown education :", unknown_education)
print("Unknown default :", unknown_default)
print("Unknown housing :", unknown_housing)
print("Unknown loans :", unknown_loans)
print("Unknown pdays :", unknown_pday)
print("Unknown pouts :", unknown_pouts)


Unknown values

Unknown jobs : 330
Unknown marital status : 80
Unknown education : 1731
Unknown default : 8597
Unknown housing : 990
Unknown loans : 990
Unknown pdays : 39673
Unknown pouts : 35563


### Observations from Data Set

- Total number of rows: 41188
- Total number of attributes: 19 + output attribute (y)
- Yes samples - 4640
- No samples - 36548

- The number of yes samples are much smaller than the number of no samples indicating that the data set is highly skewed.

- Unknown values :-
    - job - 288
    - education - 1857
    - contact - 13020
    - pdays - 36954
    - poutcome - 36959
- A very large number of rows have pdays = 999 and poutcome = nonexistent. These represent clients that have not been contacted in the past

In [7]:
df = df.drop('duration', axis=1)             # Removing duration attribute

numeric = ['age', 'campaign', 'pdays', 'previous', 'emp.var.rate', 'cons.price.idx', 'cons.conf.idx', 'euribor3m', 'nr.employed']
categorical = ['job', 'marital', 'education', 'default', 'housing', 'loan', 'contact', 'month', 'day_of_week', 'poutcome']


In [8]:
# Taking care of categorical values

# factorize assigns unique numeric values to each category
#for c in categorical:
#    df[c] = pd.factorize(df[c])[0]
    
# Use dummy attributes for categorical values
df = pd.get_dummies(df, columns=categorical, prefix=categorical)
df['y'],un = pd.factorize(df['y'])

#df[numeric] = df[numeric].astype('float')
#df[categorical] = df[categorical].astype('category')

#df['y'] = df['y'].astype(int) # Converting otuput variable to integers

In [9]:
# Normalizing numerical values

scaler = preprocessing.StandardScaler()
rescaled_df = scaler.fit_transform(df[numeric])
df[numeric] = pd.DataFrame(rescaled_df, columns=numeric)

In [10]:
#Saving as pickle file

df.to_pickle('data/bank_additional_data_dummies.pkl') # using dummies and replacing pdays only
#df.to_pickle('data/bank_additional_data_factorize.pkl') # using dummies and replacing pdays only

In [74]:
# Principal Component Analysis

y_values = df.loc[:, ['y']].values
df.drop(columns=['y'], axis=1, inplace=True)

#pca = PCA()
pca = PCA(n_components=10)
components = pca.fit_transform(df)
var_ratios = pca.explained_variance_ratio_
#rint(var_ratios)

# Calculating the number of attributes for 80% variance
s=0
count = 0
for ratio in var_ratios:
    s = s + ratio
    count = count + 1
    if s >= 0.8:
        break
print(count)

# Calculating the number of attributes for 95% variance
s=0
count1 = 0
for ratio in var_ratios:
    s = s + ratio
    count1 = count1+ 1
    if s >= 0.95:
        break
print(count1)

df = pd.DataFrame(data=components)
df['y'] = y_values

10
10


### Observations from PCA

- There were 18 attributes (after removing duration) subjected to PCA
- 80% of the variance can be captured by 6 attributes after PCA
- 95% of the variance can be captured by 10 attributes after PCA


In [75]:
df

0         1         2         3         4         5         6  \
0     -1.410808  1.044855 -1.504839 -0.035479 -0.674420  0.942602  0.541676   
1     -1.492236  1.062992 -1.537888  0.062077 -0.800902  1.043149  0.556289   
2     -1.359959  0.434188 -0.129926 -0.921168 -0.060613  0.852819 -0.717362   
3     -1.392443  0.580200 -0.350919 -0.819922 -0.098910  0.966864  0.573939   
4     -1.401831  0.979216 -1.358230 -0.096896 -0.633943  0.944694  0.465331   
5     -1.497946  0.744891 -0.782792 -0.430506 -0.482862  1.122621  0.571079   
6     -1.397250  1.094465 -1.608873 -0.010656 -0.637127  0.827676  0.542967   
7     -1.503121  0.653115 -0.550704 -0.575031 -0.395866  1.173157  0.573862   
8     -1.315879  0.013649  0.969964 -1.719263  0.567090  0.783060 -0.751126   
9     -1.306606  0.004941  0.976022 -1.650575  0.484451  0.897293 -0.745559   
10    -1.503121  0.653115 -0.550704 -0.575031 -0.395866  1.173157  0.573862   
11    -1.306606  0.004941  0.976022 -1.650575  0.484451  0.897293 -0.745559   
12    -1.356348  0.138525  0.726908 -1.440776  0.294096  1.132119  0.461007   
13    -1.354280  0.985329 -1.467035 -0.049841 -0.630251  0.696113 -0.776960   
14    -1.379349  0.385839 -0.059881 -0.947827 -0.081831  0.916459 -0.718722   
15    -1.451450  0.975151 -1.467998  0.007498 -0.747726  0.822037 -0.842825   
16    -1.379349  0.385839 -0.059881 -0.947827 -0.081831  0.916459 -0.718722   
17    -1.472796  0.723378 -0.881175 -0.345577 -0.543440  0.919690 -0.825939   
18    -1.386255  0.765813 -1.024254 -0.286244 -0.546471  0.841086 -0.828271   
19    -1.441528  0.490469 -0.109063 -0.894967 -0.108707  1.102333  0.533900   
20    -1.389019  0.305927  0.325224 -1.232184  0.149432  1.086356  0.588301   
21    -1.477907  0.994894 -1.536847  0.078508 -0.839497  0.901450 -0.753227   
22    -1.302779  0.846792 -1.087096 -0.310316 -0.415510  0.719851 -0.795905   
23    -1.315226  0.444291 -0.093852 -0.987865  0.054132  0.740398 -0.765600   
24    -1.348249  0.439614 -0.116777 -0.967850  0.018454  0.778772 -0.717225   
25    -1.388373  0.450273 -0.032655 -1.086908  0.122838  0.854381  0.488601   
26    -1.373833  1.061962 -1.653366  0.022976 -0.676780  0.645673 -0.753801   
27    -1.493264  0.594822 -0.410498 -0.691704 -0.291762  1.131539  0.577486   
28    -1.469795  0.906799 -1.022172 -0.091373 -0.154212  0.847063  0.510693   
29    -1.462043  1.021886 -1.476730 -0.065379 -0.639110  0.801649 -0.148646   
...         ...       ...       ...       ...       ...       ...       ...   
41158  1.791686 -1.053921  1.291316  0.674737 -1.767747  1.691240  0.180226   
41159  5.572664  6.041244  4.942049  1.211149 -2.790231 -0.755919 -0.253064   
41160  1.806321 -1.093715  1.392962  0.530578 -1.604076  1.597161  0.198035   
41161  2.628833 -0.064714  2.035282  0.731929 -1.908908  0.961036 -0.419652   
41162  3.674599  3.927889  1.596186  2.259867 -2.746653  0.265778 -1.698987   
41163  4.288302  4.182009  3.972002  1.448160 -1.837610 -0.082275 -1.200920   
41164  3.661416  3.791822  2.070511  1.856712 -2.383148  0.333157 -0.381338   
41165  1.779142 -0.996712  1.217875  0.995971 -1.527301  1.609363  0.178348   
41166  2.366348  0.133953  2.182218  0.640924 -1.911622  2.013456  0.782430   
41167  1.772279 -1.171209  1.586350  1.015123 -1.093981  1.411690 -1.094146   
41168  1.759314 -0.963154  1.105800  1.059374 -1.594587  1.636390  0.192450   
41169  1.671184 -0.392840 -0.236082  2.936096 -1.680470  1.475761 -1.122288   
41170  4.436396  2.898425  3.376319  1.715778 -2.612104 -0.036220  0.981779   
41171  1.608486 -0.959312  1.422570  0.597806 -1.750367  2.335163 -1.214169   
41172  1.898868 -1.320853  1.848348  0.259994 -1.347323  1.350538 -1.146304   
41173  3.238127  1.663412  0.600122  2.238508 -3.100226  0.510006  0.005387   
41174  6.816309  8.693961  4.170153  2.717164 -4.058649 -1.637012  0.654932   
41175  3.283656  0.786136  2.832299  0.825743 -2.141954  0.759755  0.003773   
41176  1.846833 -1.079965  1.198990  0.810699 -1.839889  1.46756

In [76]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 41188 entries, 0 to 41187
Data columns (total 11 columns):
0    41188 non-null float64
1    41188 non-null float64
2    41188 non-null float64
3    41188 non-null float64
4    41188 non-null float64
5    41188 non-null float64
6    41188 non-null float64
7    41188 non-null float64
8    41188 non-null float64
9    41188 non-null float64
y    41188 non-null int64
dtypes: float64(10), int64(1)
memory usage: 3.5 MB


In [77]:
df.describe()

0             1             2             3             4  \
count  4.118800e+04  4.118800e+04  4.118800e+04  4.118800e+04  4.118800e+04   
mean  -5.465183e-16  1.849329e-16  8.004560e-17  9.798686e-17 -4.278299e-17   
std    2.028186e+00  1.196995e+00  1.080178e+00  1.002392e+00  9.596998e-01   
min   -3.364377e+00 -2.555953e+00 -6.489687e+00 -2.721793e+00 -4.058649e+00   
25%   -1.475767e+00 -6.181178e-01 -6.143786e-01 -6.234059e-01 -6.071674e-01   
50%   -1.293096e+00  4.766450e-02  4.366873e-02 -9.751535e-02 -1.317384e-01   
75%    2.087327e+00  4.749275e-01  6.633857e-01  4.720083e-01  4.486139e-01   
max    7.426492e+00  9.846715e+00  6.705459e+00  1.334327e+01  1.308671e+01   

                  5             6             7             8             9  \
count  4.118800e+04  4.118800e+04  4.118800e+04  4.118800e+04  4.118800e+04   
mean  -4.278299e-17  1.359395e-16  1.826903e-16 -1.255888e-16  2.027017e-16   
std    8.361479e-01  6.926512e-01  6.838190e-01  6.371992e-01  5.555453e-01   
min   -2.691589e+00 -1.866250e+00 -2.916052e+00 -2.632857e+00 -1.266601e+00   
25%   -8.306479e-01 -6.684158e-01 -3.556995e-01 -4.445745e-01 -4.776802e-01   
50%    5.813364e-02 -4.308235e-02 -5.638620e-02 -2.330935e-02  6.292974e-02   
75%    7.608282e-01  6.163890e-01  2.463296e-01  4.325032e-01  4.106837e-01   
max    2.893686e+00  2.852630e+00  7.211624e+00  2.820911e+00  1.802491e+00   

                  y  
count  41188.000000  
mean       0.112654  
std        0.316173  
min        0.000000  
25%        0.000000  
50%        0.000000  
75%        0.000000  
max        1.000000

In [78]:
#Saving as pickle file

df.to_pickle('data/bank_additional_data_dummies_pca.pkl')    # using dummies and replacing pdays only and performing PCA